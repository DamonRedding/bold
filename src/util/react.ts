import { MutableRefObject, Ref } from 'react'

export function composeRefs<T>(...refs: Array<Ref<T>>) {
    return (value: T) => refs.forEach(ref => setRef(ref, value))
}

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}

/**
 * Returns the next sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getNextSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.nextElementSibling) {
        current = current.nextElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}

/**
 * Returns the previous sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getPreviousSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.previousElementSibling) {
        current = current.previousElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}
