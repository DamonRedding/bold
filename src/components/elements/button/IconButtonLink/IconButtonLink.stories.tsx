import { text, withKnobs } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import * as React from 'react'
import { MemoryRouter } from 'react-router'

import { withPropTypes } from '../../../../stories-addons/withPropTypes'
import { withTheme } from '../../../../stories-addons/withTheme'

import { IconButtonLink } from './IconButtonLink'

storiesOf('Components', module)
    .addDecorator(withPropTypes(`
    Botão que renderiza um link com a tag <a> ao invés de um <button>.
    `))
    .addDecorator(withKnobs)
    .addDecorator(withTheme())
    .add('IconButtonLink', () => (
        <MemoryRouter initialEntries={['/']} initialIndex={0}>
            <IconButtonLink
                to={text('to', '/')}
                icon='pen'
            />
        </MemoryRouter>
    ))
