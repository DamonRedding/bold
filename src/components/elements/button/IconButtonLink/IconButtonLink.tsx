import * as React from 'react'
import { Link, LinkProps } from 'react-router-dom'

import { withStyles, WithStylesProps } from '../../../../styles'
import { Icons } from '../../Icon/generated/Icons'
import { Icon } from '../../Icon/Icon'
import { createStyles } from '../IconButton/IconButton'

export interface IconButtonLinkProps extends WithStylesProps, Pick<LinkProps, 'to' | 'replace'> {
    icon: Icons
}

@withStyles
export class IconButtonLink extends React.PureComponent<IconButtonLinkProps> {

    render() {
        const { css, icon, theme, ...rest } = this.props
        const styles = {
            ...createStyles(theme),
            link: {
                textDecoration: 'none',
            },
        }

        const classes = css(
            styles.link,
            styles.button
        )

        return (
            <Link className={classes} {...rest}>
                <Icon icon={icon} />
            </Link>
        )
    }
}
