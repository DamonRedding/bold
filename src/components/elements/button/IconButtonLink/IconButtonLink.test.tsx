import { mount, render, shallow } from 'enzyme'
import * as React from 'react'
import { MemoryRouter } from 'react-router'

import { withTheme } from '../../../../test'

import { IconButtonLink } from './IconButtonLink'

it('deve renderizar corretamente', () => {
    const wrapper = render(withTheme(
        <MemoryRouter initialEntries={['/']} initialIndex={0} >
            <IconButtonLink to='/' icon='pen' />
        </MemoryRouter >
    ))
    expect(wrapper).toMatchSnapshot()
})
