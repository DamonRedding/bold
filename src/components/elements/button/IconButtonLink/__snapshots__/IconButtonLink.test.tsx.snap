// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deve renderizar corretamente 1`] = `
.emotion-2 {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: transparent;
  border: none;
  border-radius: 4px;
  color: #666666;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 0;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all .2s;
  transition: all .2s;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
}

.emotion-2 span {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-transition: color .2s;
  transition: color .2s;
}

.emotion-2:disabled {
  opacity: 0.5;
}

.emotion-2:focus {
  outline: none;
  box-shadow: 0 0 0 2px #056DFF;
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-0 {
  fill: currentColor;
  font-size: 1.5rem;
}

<a
  class="emotion-2"
  href="/"
>
  <span
    class="emotion-1"
  >
    <svg
      aria-hidden="true"
      class="emotion-0"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <g
        id="Editar"
      >
        <path
          d="M4 17.18L4.01 20h2.83L18.01 8.83 15.18 6 4 17.18zM19.72 5.61l-1.33-1.33a.94.94 0 0 0-1.33 0L16 5.33 18.67 8l1-1.06a.94.94 0 0 0 .05-1.33z"
        />
      </g>
    </svg>
  </span>
</a>
`;
